File: tooling/importers/dedupe.mjs
// Dedupe loader (KB 09). Merges points that are within ~30 m AND have similar names
// (likely the same place from different sources). Idempotent and re-runnable.
// Dry-run by default — pass --apply to actually merge.
//   node --env-file=.env tooling/importers/dedupe.mjs [--apply]
import postgres from 'postgres';

const APPLY = process.argv.includes('--apply');
const RADIUS_M = Number(process.env.DEDUPE_RADIUS_M || 30);
const NAME_SIM = Number(process.env.DEDUPE_NAME_SIM || 0.6);

const sql = postgres(process.env.DATABASE_URL, { ssl: 'require', max: 1, prepare: false });

function normalize(s) {
  return (s || '')
    .toLowerCase()
    .replace(/[«»"'’.,()]/g, ' ')
    .replace(/\s+/g, ' ')
    .trim();
}
function tokens(s) {
  return new Set(normalize(s).split(' ').filter(Boolean));
}
function jaccard(a, b) {
  const A = tokens(a);
  const B = tokens(b);
  if (!A.size || !B.size) return 0;
  let inter = 0;
  for (const x of A) if (B.has(x)) inter++;
  return inter / (A.size + B.size - inter);
}
function meters(a, b) {
  const R = 6371000;
  const dLat = ((b.lat - a.lat) * Math.PI) / 180;
  const dLng = ((b.lng - a.lng) * Math.PI) / 180;
  const lat = ((a.lat + b.lat) / 2) * (Math.PI / 180);
  const x = dLng * Math.cos(lat);
  return Math.sqrt(x * x + dLat * dLat) * R;
}

// union-find
function find(parent, i) { while (parent[i] !== i) { parent[i] = parent[parent[i]]; i = parent[i]; } return i; }
function union(parent, i, j) { parent[find(parent, i)] = find(parent, j); }

async function main() {
  const rows = await sql`
    select p.id, p.name, p.category, p.verify_status, p.created_at, p.photos,
           ST_X(p.geom::geometry) as lng, ST_Y(p.geom::geometry) as lat,
           (select count(*) from point_feature_values fv where fv.point_id = p.id)::int as nfeatures
    from points p order by p.created_at`;
  console.log(`Scanning ${rows.length} points (radius ${RADIUS_M} m, name-sim ≥ ${NAME_SIM}) …`);

  const parent = rows.map((_, i) => i);
  for (let i = 0; i < rows.length; i++) {
    for (let j = i + 1; j < rows.length; j++) {
      if (rows[i].category !== rows[j].category) continue;
      if (meters(rows[i], rows[j]) > RADIUS_M) continue;
      if (jaccard(rows[i].name, rows[j].name) < NAME_SIM) continue;
      union(parent, i, j);
    }
  }

  const clusters = new Map();
  for (let i = 0; i < rows.length; i++) {
    const r = find(parent, i);
    if (!clusters.has(r)) clusters.set(r, []);
    clusters.get(r).push(i);
  }

  const score = (r) => r.nfeatures * 10 + (r.verify_status === 'official' ? 5 : r.verify_status === 'verified' ? 3 : 0);
  let merged = 0;
  for (const idxs of clusters.values()) {
    if (idxs.length < 2) continue;
    const members = idxs.map((i) => rows[i]).sort((a, b) => score(b) - score(a) || new Date(a.created_at) - new Date(b.created_at));
    const survivor = members[0];
    const dups = members.slice(1);
    console.log(`MERGE «${survivor.name}» ← ${dups.map((d) => `«${d.name}»`).join(', ')}`);
    merged += dups.length;
    if (!APPLY) continue;

    for (const dup of dups) {
      // Move feature values the survivor doesn't already have, then drop the rest.
      await sql`
        update point_feature_values fv set point_id = ${survivor.id}
        where fv.point_id = ${dup.id}
          and not exists (select 1 from point_feature_values s where s.point_id = ${survivor.id} and s.feature_key = fv.feature_key)`;
      await sql`delete from point_feature_values where point_id = ${dup.id}`;
      // Repoint reviews the survivor doesn't already have for that user/profile.
      await sql`
        update reviews r set point_id = ${survivor.id}
        where r.point_id = ${dup.id}
          and not exists (select 1 from reviews s where s.point_id = ${survivor.id} and s.user_id = r.user_id and s.profile = r.profile)`;
      await sql`delete from reviews where point_id = ${dup.id}`;
      // Repoint problems.
      await sql`update problems set point_id = ${survivor.id} where point_id = ${dup.id}`;
      // Merge photos.
      const photos = Array.from(new Set([...(survivor.photos || []), ...(dup.photos || [])]));
      await sql`update points set photos = ${photos} where id = ${survivor.id}`;
      await sql`delete from points where id = ${dup.id}`;
    }
  }

  console.log(APPLY ? `Dedupe applied: merged ${merged} duplicate(s).` : `Dry run: ${merged} duplicate(s) would be merged. Re-run with --apply.`);
}

try {
  await main();
} catch (e) {
  console.error('dedupe failed:', e.message);
  process.exitCode = 1;
} finally {
  await sql.end();
}

File: apps/api/internal/importer/dedupe.go
package importer

import (
	"context"
	"fmt"
	"math"
	"regexp"
	"sort"
	"strings"
	"time"

	"github.com/jackc/pgx/v5/pgxpool"
)

// Dedupe defaults (overridable via the CLI / env).
const (
	DefaultRadiusM = 30.0
	DefaultNameSim = 0.6
)

// DedupeRow is a point considered for merging.
type DedupeRow struct {
	ID           string
	Name         string
	Category     string
	VerifyStatus string
	CreatedAt    time.Time
	Photos       []string
	Lng, Lat     float64
	NFeatures    int
}

// Merge is a survivor and the duplicates folded into it.
type Merge struct {
	Survivor DedupeRow
	Dups     []DedupeRow
}

// DedupeReport summarizes a dedupe run.
type DedupeReport struct {
	Scanned int
	Merges  []Merge
	Applied bool
}

// MergedCount is the total number of duplicate rows folded away.
func (r DedupeReport) MergedCount() int {
	n := 0
	for _, m := range r.Merges {
		n += len(m.Dups)
	}
	return n
}

var punctRe = regexp.MustCompile(`[«»"'’.,()]+`)
var spaceRe = regexp.MustCompile(`\s+`)

func normalizeName(s string) string {
	s = strings.ToLower(s)
	s = punctRe.ReplaceAllString(s, " ")
	s = spaceRe.ReplaceAllString(s, " ")
	return strings.TrimSpace(s)
}

func nameTokens(s string) map[string]struct{} {
	set := map[string]struct{}{}
	for _, tok := range strings.Fields(normalizeName(s)) {
		set[tok] = struct{}{}
	}
	return set
}

// jaccard is the token-set Jaccard similarity of two names.
func jaccard(a, b string) float64 {
	A, B := nameTokens(a), nameTokens(b)
	if len(A) == 0 || len(B) == 0 {
		return 0
	}
	inter := 0
	for x := range A {
		if _, ok := B[x]; ok {
			inter++
		}
	}
	return float64(inter) / float64(len(A)+len(B)-inter)
}

// metersBetween is an equirectangular-approx distance in metres.
func metersBetween(aLat, aLng, bLat, bLng float64) float64 {
	const R = 6371000
	dLat := (bLat - aLat) * math.Pi / 180
	dLng := (bLng - aLng) * math.Pi / 180
	lat := (aLat + bLat) / 2 * math.Pi / 180
	x := dLng * math.Cos(lat)
	return math.Sqrt(x*x+dLat*dLat) * R
}

func ufFind(parent []int, i int) int {
	for parent[i] != i {
		parent[i] = parent[parent[i]]
		i = parent[i]
	}
	return i
}

func ufUnion(parent []int, i, j int) { parent[ufFind(parent, i)] = ufFind(parent, j) }

func score(r DedupeRow) int {
	s := r.NFeatures * 10
	switch r.VerifyStatus {
	case "official":
		s += 5
	case "verified":
		s += 3
	}
	return s
}

// clusterDuplicates groups rows that share a category, sit within radiusM, and
// have name similarity ≥ nameSim, returning one Merge per cluster of size ≥ 2.
// The survivor is the highest-scoring (most features, most trusted), oldest first.
func clusterDuplicates(rows []DedupeRow, radiusM, nameSim float64) []Merge {
	parent := make([]int, len(rows))
	for i := range parent {
		parent[i] = i
	}
	for i := 0; i < len(rows); i++ {
		for j := i + 1; j < len(rows); j++ {
			if rows[i].Category != rows[j].Category {
				continue
			}
			if metersBetween(rows[i].Lat, rows[i].Lng, rows[j].Lat, rows[j].Lng) > radiusM {
				continue
			}
			if jaccard(rows[i].Name, rows[j].Name) < nameSim {
				continue
			}
			ufUnion(parent, i, j)
		}
	}

	groups := map[int][]int{}
	for i := range rows {
		root := ufFind(parent, i)
		groups[root] = append(groups[root], i)
	}

	// Deterministic order: by smallest member index.
	roots := make([]int, 0, len(groups))
	for r := range groups {
		roots = append(roots, r)
	}
	sort.Ints(roots)

	var merges []Merge
	for _, r := range roots {
		idxs := groups[r]
		if len(idxs) < 2 {
			continue
		}
		members := make([]DedupeRow, len(idxs))
		for k, i := range idxs {
			members[k] = rows[i]
		}
		sort.SliceStable(members, func(a, b int) bool {
			if sa, sb := score(members[a]), score(members[b]); sa != sb {
				return sa > sb
			}
			return members[a].CreatedAt.Before(members[b].CreatedAt)
		})
		merges = append(merges, Merge{Survivor: members[0], Dups: members[1:]})
	}
	return merges
}

const dedupeLoadSQL = `
select p.id::text, p.name, p.category::text, p.verify_status::text, p.created_at, p.photos,
       ST_X(p.geom::geometry) as lng, ST_Y(p.geom::geometry) as lat,
       (select count(*) from point_feature_values fv where fv.point_id = p.id)::int as nfeatures
from points p order by p.created_at`

// Dedupe scans all points and merges proximity+name duplicates. Dry-run unless
// apply is true.
func Dedupe(ctx context.Context, pool *pgxpool.Pool, radiusM, nameSim float64, apply bool) (DedupeReport, error) {
	rows, err := loadDedupeRows(ctx, pool)
	if err != nil {
		return DedupeReport{}, err
	}
	report := DedupeReport{Scanned: len(rows), Merges: clusterDuplicates(rows, radiusM, nameSim), Applied: apply}
	if !apply {
		return report, nil
	}
	for _, m := range report.Merges {
		if err := applyMerge(ctx, pool, m); err != nil {
			return report, fmt.Errorf("merge into %s: %w", m.Survivor.ID, err)
		}
	}
	return report, nil
}

func loadDedupeRows(ctx context.Context, pool *pgxpool.Pool) ([]DedupeRow, error) {
	rows, err := pool.Query(ctx, dedupeLoadSQL)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []DedupeRow
	for rows.Next() {
		var r DedupeRow
		if err := rows.Scan(&r.ID, &r.Name, &r.Category, &r.VerifyStatus, &r.CreatedAt,
			&r.Photos, &r.Lng, &r.Lat, &r.NFeatures); err != nil {
			return nil, err
		}
		out = append(out, r)
	}
	return out, rows.Err()
}

// applyMerge folds each duplicate into the survivor inside one transaction.
func applyMerge(ctx context.Context, pool *pgxpool.Pool, m Merge) (err error) {
	tx, err := pool.Begin(ctx)
	if err != nil {
		return err
	}
	defer func() {
		if err != nil {
			_ = tx.Rollback(ctx)
		}
	}()

	photos := append([]string{}, m.Survivor.Photos...)
	for _, dup := range m.Dups {
		// Move feature values the survivor lacks, then drop the rest.
		if _, err = tx.Exec(ctx, `
			update point_feature_values fv set point_id = $1
			where fv.point_id = $2
			  and not exists (select 1 from point_feature_values s
			                  where s.point_id = $1 and s.feature_key = fv.feature_key)`,
			m.Survivor.ID, dup.ID); err != nil {
			return err
		}
		if _, err = tx.Exec(ctx, `delete from point_feature_values where point_id = $1`, dup.ID); err != nil {
			return err
		}
		// Repoint reviews the survivor lacks for that user/profile, drop the rest.
		if _, err = tx.Exec(ctx, `
			update reviews r set point_id = $1
			where r.point_id = $2
			  and not exists (select 1 from reviews s
			                  where s.point_id = $1 and s.user_id = r.user_id and s.profile = r.profile)`,
			m.Survivor.ID, dup.ID); err != nil {
			return err
		}
		if _, err = tx.Exec(ctx, `delete from reviews where point_id = $1`, dup.ID); err != nil {
			return err
		}
		// Repoint problems.
		if _, err = tx.Exec(ctx, `update problems set point_id = $1 where point_id = $2`,
			m.Survivor.ID, dup.ID); err != nil {
			return err
		}
		photos = mergePhotos(photos, dup.Photos)
		if _, err = tx.Exec(ctx, `delete from points where id = $1`, dup.ID); err != nil {
			return err
		}
	}
	if _, err = tx.Exec(ctx, `update points set photos = $1 where id = $2`, photos, m.Survivor.ID); err != nil {
		return err
	}
	return tx.Commit(ctx)
}

func mergePhotos(a, b []string) []string {
	seen := map[string]struct{}{}
	out := []string{}
	for _, p := range append(append([]string{}, a...), b...) {
		if _, ok := seen[p]; !ok {
			seen[p] = struct{}{}
			out = append(out, p)
		}
	}
	return out
}
